<html> <h1 style="font-style:bold; color:blue;"> Machine Learning in Finance </h1> </html>

<html> <h1 style="font-style:italic; color:blue;"> Week-1 </h1> </html>


# NumPy

The main python library used in numerical data analysis is NumPy (or Numpy). It is a Linear Algebra Library for Python.  It is a very important library for Data Science with Python mainly because Numpy is incredibly fast, as it has bindings to C libraries.

It provides a high-performance multidimensional array object, and tools for working with these arrays.

This practical gives only an introduction to Numpy.

## Installation NumPy

**In the case of using the Anaconda distribution, the Numpay has already been installed. If you have any problem with Numpay functionality, it is highly recommended to re-install Numpay using the following command:**
    
    install numpy


____________________________________________![image.png](attachment:d7dc4e19-f52e-4a93-b2f1-e4cd231d4be9.png)

In [10]:
import numpy as np

One of the most popular features of the Numpy library is the capability of operations with arrays of different sizes (or ranks).

Dimensions of arrays:
- vectors - One dimensional arrays,
- matrices - Two dimensional arrays,
- tensor - Three or more arrays.

49.png![image.png](attachment:image.png)

____________________________________________![image.png](attachment:c8b0e131-8ac3-4605-b402-9cca1b5ba9d3.png)

In [12]:
a=np.array([1,2,3])

____________________________________________![image.png](attachment:565f7ffe-ea75-4f62-b130-093f60e7e041.png)

In [13]:
print(type(a))
print(a.shape)

<class 'numpy.ndarray'>
(3,)


____________________________________________![image.png](attachment:cd442ab7-487e-4b31-ab69-0177198e683e.png)

In [14]:
print(a[0],a[1],a[2])

1 2 3


1 2 3
[1 2 3]


____________________________________________![image.png](attachment:8acae197-093e-403c-adaa-e2eefa1cca1a.png)

In [15]:
a[0]=5
print(a)

[5 2 3]


[5 2 3]


____________________________________________![image.png](attachment:f676ad92-bb1d-49c6-bff6-e4aca2174453.png)

In [16]:
b=np.array([[1,2,3],[4,5,6]])
b

array([[1, 2, 3],
       [4, 5, 6]])

array([[1, 2, 3],
       [4, 5, 6]])

____________________________________________![image.png](attachment:193cd9e0-7861-4552-9ac9-237f88534427.png)

In [17]:
print(b[1,0])

4


4


### arange

The arange() method is used to return an evenly spaced values within a given interval.
The method has the following parameters:
- **start** parameter (inclusive in the sequence)  
- **end** parameter (exclusive in the sequence)
- **step size** (optional parameter used to find the next element in the sequence)


____________________________________________![image.png](attachment:be72cd71-324f-4b2f-a9eb-9f58864ee035.png)

In [18]:
np.arange(10,20)

array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19])

array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19])

____________________________________________![image.png](attachment:793515fa-88f9-471d-bb3d-6b28e65a20c1.png)

In [19]:
np.arange(30,1,-3)

array([30, 27, 24, 21, 18, 15, 12,  9,  6,  3])

array([30, 27, 24, 21, 18, 15, 12,  9,  6,  3])

<html> <h3 style="font-style:italic; color:blue;"> Let us compare the performance of Numpay and Python arrays. </h3> </html>

In [ ]:
# timeit - executes the code several times and gives statistics.

#### Now just press 'CTRL' and 'ENTER' keys on the next cells.

In [20]:
%timeit np.arange(0, 1000)

1.83 µs ± 501 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [ ]:
%timeit range(0, 1000)

52.9 ns ± 14.3 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


<html> <h3 style="color:blue;"> Numpy also provides many functions to create arrays </h3> </html>

____________________________________________![image.png](attachment:5644d1a1-af6f-499a-8b57-56e2b163cf07.png)

In [21]:
a=np.zeros((2,2))
a

array([[0., 0.],
       [0., 0.]])

array([[0., 0.],
       [0., 0.]])

____________________________________________![image.png](attachment:2b89ec2a-047e-4e58-9ca8-fcb0f4cb6b5b.png)

In [22]:
b=np.ones((3,4))
b

array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

____________________________________________![image.png](attachment:267a89d5-8030-4262-b0b8-6c744458ad89.png)

In [23]:
np.eye(4)

array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]])

array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]])

____________________________________________![image.png](attachment:339d2c8c-51d3-4874-9dd6-ccc57337eeb1.png)

In [24]:
e=np.random.random((2,2))
e

array([[0.52676268, 0.91515531],
       [0.11793938, 0.28676903]])

array([[0.19249862, 0.99025753],
       [0.15139398, 0.09978187]])

____________________________________________![image.png](attachment:155cb7d7-48d9-4b79-a551-503d42cca2b9.png)

In [25]:
np.linspace(0,20,11)

array([ 0.,  2.,  4.,  6.,  8., 10., 12., 14., 16., 18., 20.])

array([ 0.,  2.,  4.,  6.,  8., 10., 12., 14., 16., 18., 20.])

____________________________________________![image.png](attachment:46117baf-57b3-493c-b73d-d3255b181220.png)

In [26]:
np.linspace(0,20,30)

array([ 0.        ,  0.68965517,  1.37931034,  2.06896552,  2.75862069,
        3.44827586,  4.13793103,  4.82758621,  5.51724138,  6.20689655,
        6.89655172,  7.5862069 ,  8.27586207,  8.96551724,  9.65517241,
       10.34482759, 11.03448276, 11.72413793, 12.4137931 , 13.10344828,
       13.79310345, 14.48275862, 15.17241379, 15.86206897, 16.55172414,
       17.24137931, 17.93103448, 18.62068966, 19.31034483, 20.        ])

array([ 0.        ,  0.68965517,  1.37931034,  2.06896552,  2.75862069,
        3.44827586,  4.13793103,  4.82758621,  5.51724138,  6.20689655,
        6.89655172,  7.5862069 ,  8.27586207,  8.96551724,  9.65517241,
       10.34482759, 11.03448276, 11.72413793, 12.4137931 , 13.10344828,
       13.79310345, 14.48275862, 15.17241379, 15.86206897, 16.55172414,
       17.24137931, 17.93103448, 18.62068966, 19.31034483, 20.        ])

<html> <h3 style="color:blue;"> Array Indexing </h3> </html>

**Slicing:**
    Similar to Python lists, numpy arrays can be sliced. Since arrays may be multidimensional, you must specify a slice for each dimension of the array:

..................................................![image.png](attachment:11cabfa7-9787-4a3a-94d5-f27b5089bbea.png)

In [27]:
a=np.array([[1,2,3,4],[5,6,7,8],[9,10,11,12]])
a

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])

____________________________________________![image.png](attachment:b4086256-1bbb-4e1c-820c-415d37eb11b2.png)

In [28]:
b=a[:2,1:3]
b

array([[2, 3],
       [6, 7]])

array([[2, 3],
       [6, 7]])

____________________________________________![image.png](attachment:f3b12e40-eda9-467d-9711-d9b57f20f6ea.png)

In [29]:
print(a[0,1])

2


2


____________________________________________![image.png](attachment:3bf651f2-ee5c-4cf3-a41c-8bbfd13bd70f.png)

In [30]:
b[0,0]=88
print(a)

[[ 1 88  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]


<html> <h3 style="color:blue;"> Datatypes </h3> </html>

Every NumPy array is a grid of elements of the **same type**. Numpy provides a large set of numeric datatypes that you can use to construct arrays. Numpy tries to guess a datatype when you create an array, but you can explicitly specify the type using an optional argument.

____________________________________________![image.png](attachment:acf9492f-76d6-451f-90d6-f8af21190883.png)

In [31]:
x=np.array([1,2])
print(x.dtype)

int64


int32


____________________________________________![image.png](attachment:abd5f58f-2ced-436f-bf31-f5455975721d.png)

In [32]:
x = np.array([1.0, 2.0])
print(x.dtype)

float64


float64


____________________________________________![image.png](attachment:cc855679-b5ab-4be8-859f-cbecf8a0fbba.png)

In [33]:
x=np.array([1,2], dtype=np.int64)
print(x.dtype)

int64


int64


<html> <h3 style="color:blue;"> Array Math </h3> </html>

In [34]:
# Basic mathematical functions operate elementwise on arrays

.....................................................![image.png](attachment:62b9a0d9-3d99-4b14-a8ad-3711111afd5f.png)

#### Now just press 'CTRL' and 'ENTER' keys on the next cell.

In [35]:
x = np.array([[1,2],[3,4]], dtype=np.float64)
y = np.array([[5,6],[7,8]], dtype=np.float64)
print(x, '\n')       # '\n' - Line break symbol
print(y)

[[1. 2.]
 [3. 4.]] 

[[5. 6.]
 [7. 8.]]


[[1. 2.]
 [3. 4.]] 

[[5. 6.]
 [7. 8.]]


____________________________________________![image.png](attachment:c0c48780-516e-48fc-93be-796adc25d912.png)

In [36]:
print(x+y)
print('\n')
print(np.add(x,y))

[[ 6.  8.]
 [10. 12.]]


[[ 6.  8.]
 [10. 12.]]


[[ 6.  8.]
 [10. 12.]]


[[ 6.  8.]
 [10. 12.]]


____________________________________________![image.png](attachment:07714d60-442c-4abd-b584-a8d7af18d9bb.png)

In [40]:
print(x-y, '\n')
print(np.subtract(x,y))

[[-4. -4.]
 [-4. -4.]] 

[[-4. -4.]
 [-4. -4.]]


[[-4. -4.]
 [-4. -4.]] 

[[-4. -4.]
 [-4. -4.]]


____________________________________________![image.png](attachment:b6a795f3-e338-490f-a256-39a511082445.png)

In [41]:
print(x*y,'\n')
print(np.multiply(x,y))

[[ 5. 12.]
 [21. 32.]] 

[[ 5. 12.]
 [21. 32.]]


[[ 5. 12.]
 [21. 32.]] 

[[ 5. 12.]
 [21. 32.]]


____________________________________________![image.png](attachment:3ad4361b-8759-4cf4-acda-29ef1b7d6ffc.png)

In [42]:
print(x/y)
print('\n')
print(np.divide(x,y))

[[0.2        0.33333333]
 [0.42857143 0.5       ]]


[[0.2        0.33333333]
 [0.42857143 0.5       ]]


[[0.2        0.33333333]
 [0.42857143 0.5       ]]


[[0.2        0.33333333]
 [0.42857143 0.5       ]]


____________________________________________![image.png](attachment:a6875e9c-4f66-42f3-a098-5d810431a65c.png)

In [43]:
print (np. sqrt(x))

[[1.         1.41421356]
 [1.73205081 2.        ]]


[[1.         1.41421356]
 [1.73205081 2.        ]]


____________________________________________![image.png](attachment:85fe41cf-7053-41c8-af99-29afb236745b.png)

In [45]:
x**3

array([[ 1.,  8.],
       [27., 64.]])

array([[ 1.,  8.],
       [27., 64.]])

____________________________________________![image.png](attachment:e098f400-88dd-4718-a254-9bad8f1d50a8.png)

In [46]:
1/x

array([[1.        , 0.5       ],
       [0.33333333, 0.25      ]])

array([[1.        , 0.5       ],
       [0.33333333, 0.25      ]])

<html> <h3 style="color:blue;"> Matrix Multiplication </h3> </html>

In [ ]:
# Note that the above was elementwise multiplication, not matrix multiplication.
# To receive the matrix multiplication, we use the dot function.

#### Now just press 'CTRL' and 'ENTER' keys on the next cell.

In [47]:
x = np.array([[1,2],[3,4]])
y = np.array([[5,6],[7,8]])

v = np.array([9,10])
w = np.array([11, 12])
print(x, '\n\n', y, '\n\n', v, '\n\n', w)

[[1 2]
 [3 4]] 

 [[5 6]
 [7 8]] 

 [ 9 10] 

 [11 12]


![image.png](attachment:4775c518-47b9-4860-8f55-196766891597.png)

____________________________________________![image.png](attachment:f04b0822-51b3-4ca0-9645-e930c515bb4b.png)

In [48]:
print (v. dot (w) )
print(np.dot(v, w))

219
219


219
219


![image.png](attachment:86e4646b-afd4-45a8-bca2-92656bfba0fb.png)

Watch this video carefully at **home** and remember how matrices are multiplied

https://www.youtube.com/watch?v=2spTnAiQg4M


____________________________________________![image.png](attachment:fd6fab27-ad50-469c-95af-88d37aa1565c.png)

In [49]:
print (x. dot(v))
print('\n')
print(np.dot(x,v))

[29 67]


[29 67]


[29 67]


[29 67]


In [ ]:
# an example of matrix multiplication:

![image.png](attachment:134eb75e-46f0-4059-a113-23d0eb1e4b03.png)

____________________________________________![image.png](attachment:88ed560c-87be-4ced-bf7e-e8f7f1c9d370.png)

In [50]:
print(x. dot (y))
print('\n')
print(np.dot(x,y))

[[19 22]
 [43 50]]


[[19 22]
 [43 50]]


[[19 22]
 [43 50]]


[[19 22]
 [43 50]]


<html> <h3 style="color:blue;"> Random number array generator </h3> </html>

____________________________________________![image.png](attachment:b66acf21-ce8f-41c8-bda0-8446e71611f6.png)

In [53]:
np.random.rand(4)

array([0.35226111, 0.74338259, 0.57061444, 0.62442767])

array([0.39530672, 0.94341135, 0.7134285 , 0.41761022])

____________________________________________![image.png](attachment:d502bc7f-c406-47cf-86b0-858b69aa0feb.png)

In [52]:
np. random.rand(4,4)

array([[0.30879778, 0.8368897 , 0.31102839, 0.98428303],
       [0.67627996, 0.80218547, 0.06069772, 0.62405383],
       [0.98512715, 0.50493318, 0.70122589, 0.92978844],
       [0.86687977, 0.23989483, 0.7342523 , 0.51809408]])

array([[0.12647714, 0.37823272, 0.7573721 , 0.11737587],
       [0.0020701 , 0.8499101 , 0.93522796, 0.60747499],
       [0.27604771, 0.85404114, 0.38664039, 0.58206828],
       [0.60229498, 0.18029854, 0.02289854, 0.41117278]])

##### Normal distribution (Gaussian distribution)

![image.png](attachment:2a8d6d99-a743-406b-af0b-bded80612de1.png)

______________________![image.png](attachment:197e75d0-c7df-4e33-bbef-34d556c86be8.png)

In [54]:
np. random. randn (4 , 4)

array([[-0.22403737,  0.11588111, -1.08347645,  1.60064006],
       [-0.60285496,  0.62414142,  0.42720826, -0.2271729 ],
       [ 0.71718755,  0.92495517, -1.16978246, -0.5564697 ],
       [-1.22281349,  0.21423292, -0.08603712,  2.48477926]])

array([[ 1.81518269, -2.4213843 , -0.59131028, -0.21626991],
       [-0.5849672 ,  1.11994527, -0.56770557, -0.82824582],
       [-0.55215755,  0.58089468, -1.41957238,  1.32710565],
       [-0.90394077, -0.10972108, -0.84841308,  0.09323056]])

____________________________________________![image.png](attachment:cd9a9f30-f9b9-42fb-8243-ddb743ee01c3.png)

In [57]:
np.random.randint(1,300)

116

32

____________________________________________![image.png](attachment:998bb937-89a8-416e-99da-0109fb725b6e.png)

In [58]:
np. random. randint( 1, 200, 15)

array([ 64, 131, 100, 128, 145, 159, 110,  87,  80,   3,  62, 150, 107,
        92, 112])

array([113, 180, 139, 142,  35, 135,   6, 120,  17,  26,   4, 154, 173,
       164,  34])

<html> <h3 style="color:blue;"> Reshape </h3> </html>

The shape (dimensions) of the array can be changed using the reshape() method
This returns an array containing the same data with a new shape.

____________________________________________![image.png](attachment:288fc1d4-3930-4e62-8ec9-63be664670c3.png)

In [59]:
p= np.arange(0,15)
p

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14])

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14])

____________________________________________![image.png](attachment:3c35d8d2-17f4-43bc-a513-155af5c09df6.png)

In [60]:
p=p.reshape(3,5)
p

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14]])

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14]])

____________________________________________![image.png](attachment:2ca2045d-3dba-4c89-aeac-8176014a3cdf.png)

In [61]:
s= np.arange(12) . reshape(3, 4)
s

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

##### Transposing a Matrix

![image.png](attachment:ae787138-a47b-4140-9bec-99a532903bbd.png)

____________________________________________![image.png](attachment:1e9d411f-d355-4233-9180-639e7659446d.png)

In [62]:
print(p.T)

[[ 0  5 10]
 [ 1  6 11]
 [ 2  7 12]
 [ 3  8 13]
 [ 4  9 14]]


[[ 0  5 10]
 [ 1  6 11]
 [ 2  7 12]
 [ 3  8 13]
 [ 4  9 14]]


____________________________________________![image.png](attachment:f5b163a3-7ce4-4100-942b-efc07e220eed.png)

The command does not create an "empty" matrix in the sense of being filled with zeros or having no values. Instead, it returns a new array with the same shape and data type as p, but without initialising the elements — meaning the array elements contain random values that happened to be in memory at the time of allocation

In [78]:
k=np.empty_like(p)
k

array([[           618679442,                    0,      132875763685040,
         9025942258489769987,      132875763363584],
       [     132875763686384, -2207582518940086624,      132875763686464,
             132875763687152, -3736469457979307384],
       [     132875763364160,      132875763687472,  5791461547098209276,
             132875763347520,                  128]])

array([[3801155, 5570652, 6619251, 7536754, 7733340],
       [7602281, 7077985, 4259932, 7340144, 6357060],
       [6357108, 4980828, 6488175, 7077985, 5505116]])

____________________________________________![image.png](attachment:4bc158d4-9ab4-45cb-8f92-48339dd3e6e5.png)

In [79]:
v=np.array([1,0,1])
print(v,'\n')

vv=np.tile(v,(4,1))
print(vv)

[1 0 1] 

[[1 0 1]
 [1 0 1]
 [1 0 1]
 [1 0 1]]


[1 0 1] 

[[1 0 1]
 [1 0 1]
 [1 0 1]
 [1 0 1]]


<html> <h3 style="color:blue;"> Functions </h3> </html>

____________________________________________![image.png](attachment:bf6cf184-57c9-4dd4-bfb2-7def73549040.png)

In [80]:
a

array([[ 1, 88,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])

array([[ 1, 88,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])

In [ ]:
# Numpy provides many useful functions for performing computations on arrays; one of the most useful is sum:

____________________________________________![image.png](attachment:a7779684-6a28-49f0-9c2f-4fab2a79a47e.png)

In [81]:
print(np.sum(a))

164


164


____________________________________________![image.png](attachment:26e7377d-c022-4e06-8ade-0f49b2759669.png)

In [83]:
print(np.sum(a, axis=0))

[ 15 104  21  24]


[ 15 104  21  24]


____________________________________________![image.png](attachment:e7c61121-ac0a-43d5-a824-f12b26e99c73.png)

In [84]:
print(np.sum(a, axis=1))

[96 26 42]


[96 26 42]


____________________________________________![image.png](attachment:0d80160c-983f-4bcf-9a16-6ce6911f9d4e.png)

In [85]:
a.max()

np.int64(88)

88

____________________________________________![image.png](attachment:16c464a1-f29f-4bd6-8de8-e86eab73c66d.png)

In [86]:
a.min()

np.int64(1)

1

____________________________________________![image.png](attachment:7ed980b3-88aa-43ee-a31d-3d664d24d492.png)

In [87]:
a.argmax()

np.int64(1)

1

____________________________________________![image.png](attachment:7ca34411-c78a-4d05-8f4b-64459220ad88.png)

In [88]:
a.argmin()

np.int64(0)

0

<html> <h3 style="color:blue;"> Shape </h3> </html>

The size of a Numpy array can be retrieved using the attribute named "shape". Please note it is not a method. Attribute does not use "()" when evoked.

____________________________________________![image.png](attachment:c1e69ea3-9843-4361-87d3-39c77f8936e4.png)

In [89]:
arr=np.arange(10)
print(arr)
print(arr.shape)

[0 1 2 3 4 5 6 7 8 9]
(10,)


[0 1 2 3 4 5 6 7 8 9]
(10,)


____________________________________________![image.png](attachment:53dde3eb-9adf-400d-9d8b-e94692a80ef5.png)

In [90]:
print(a)
print(a.shape)

[[ 1 88  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]
(3, 4)


[[ 1 88  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]
(3, 4)


The shape of an array can be changed with various commands. Note that the following three commands all return a modified array, but do not change the original array:

____________________________________________![image.png](attachment:d481818b-d0c0-4c0b-8443-2369390cc1fa.png)

In [91]:
a.ravel()

array([ 1, 88,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12])

array([ 1, 88,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12])

____________________________________________![image.png](attachment:e9a52f60-26dc-4230-97e6-b877ed8e8bba.png)

In [92]:
a.reshape(6,2)

array([[ 1, 88],
       [ 3,  4],
       [ 5,  6],
       [ 7,  8],
       [ 9, 10],
       [11, 12]])

array([[ 1, 88],
       [ 3,  4],
       [ 5,  6],
       [ 7,  8],
       [ 9, 10],
       [11, 12]])

____________________________________________![image.png](attachment:4ac28c80-ba5d-4357-a3e7-6fba7fc3d9f4.png)

In [93]:
print(a.shape)
print(a.T.shape)

(3, 4)
(4, 3)


(3, 4)
(4, 3)


Change matrix a to **2-d** array with 1 column. Notice the two sets of brackets for a 2-d array with one column.

Save it in another array.

Check the shape attribute value.

____________________________________________![image.png](attachment:a3063539-2ef1-4ae3-a392-e9baf8ab8ca3.png)

In [94]:
a3=a.reshape(12,1)
print(a3)
print('\n')
print(a3.shape)

[[ 1]
 [88]
 [ 3]
 [ 4]
 [ 5]
 [ 6]
 [ 7]
 [ 8]
 [ 9]
 [10]
 [11]
 [12]]


(12, 1)


[[ 1]
 [88]
 [ 3]
 [ 4]
 [ 5]
 [ 6]
 [ 7]
 [ 8]
 [ 9]
 [10]
 [11]
 [12]]


(12, 1)


<html> <h3 style="font-style:italic; color:#FF0000;"> (!) From the point of view of mathematics - this is a 1-d vector.
However, in terms of Numpy, it is a 2-dimensional array.
</h3> </html>

# Lab Logbook Requirement:

<html> <h3 style="font-style:italic; color:blue;">
1) Create a vector using np.arange.

Determine the number of the vector elements using the following method: Take the last two digits from your SID. It should be from 00 to 99. If this number is 10 or more, it becomes the required number of the vector elements. If it is less than 10, add 100 to your number.

For example, if your SID is 2287467, and the last two digits are 67, which is greater than 10. The required number is 67. If your SID is 2287407, and the last two digits are 07, which is less than 10. The required number is 107.

Then,

2) Change matrix a to 2-d array with 1 row. Print the array.
You should have the two sets of brackets for a 2-d array with one row.
3) Save it in another array. Print the array.
4) Check the shape attribute value.
5) Add the code and result to your Lab Logbook
</h3> </html>

<html> <h3 style="color:red;">
NOTE: DON'T FORGET TO SAVE AND BACK UP YOUR COMPLETED JUPYTER NOTEBOOK AND LAB LOGBOOK ON GITHUB OR ONEDRIVE.
</h3> </html>

In [98]:
import numpy as np
sid=2464644
sid_last_two = 44

if sid_last_two < 10:
    n = sid_last_two + 100
else:
    n = sid_last_two
print("Number of elements required:",n)
a = np.arange(n)
print("Vector a:", a)
a_2d = a.reshape(1, -1)
print("2D array a_2d:\n", a_2d)
b = a_2d.copy()
print("Copied array b:\n", b)
print("Shape of a_2d:", a_2d.shape)
print("Shape of b:", b.shape)

Number of elements required: 44
Vector a: [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43]
2D array a_2d:
 [[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
  24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43]]
Copied array b:
 [[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
  24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43]]
Shape of a_2d: (1, 44)
Shape of b: (1, 44)
